### ISSUE TITLE
overlayfs: Can't delete file moved from base layer to newly created dir even on ext4

### ISSUE THREAD

**Issue by overplumbum** [docker/docker#9572]
I'm testing overlayfs with docker  from master branch 58ce014 ( https://master.dockerproject.com/ ):
<pre>
Server version: 1.3.2-dev
Server API version: 1.16
Go version (server): go1.3.3
Git commit (server): 58ce014
</pre>
On Ubuntu:14.04 with 3.18 kernel from vivid ( http://kernel.ubuntu.com/~kernel-ppa/mainline/v3.18-vivid/ )
<pre>
Linux 3.18.0-031800-generic #201412071935 SMP Mon Dec 8 00:36:34 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
</pre>
<pre>
# docker -D info
Containers: 51
Images: 169
Storage Driver: overlay
Execution Driver: native-0.2
Kernel Version: 3.18.0-031800-generic
Operating System: Ubuntu 14.04.1 LTS
Debug mode (server): false
Debug mode (client): true
Fds: 10
Goroutines: 13
EventsListeners: 0
Init Path: /usr/local/bin/docker-1.3.2-dev
WARNING: No swap limit support
</pre>

With graph stored on EXT4 partition using overlay(fs) storage backend.

To reproduce issue one should:
  - create new dir
  - move some file from base layer to this new dir
  - remove this file from it's new location (no error message here)
  - check if it is still there (file transforms to some kind of character device)

Snippet to reproduce:
<pre>
$ docker run busybox sh -xec 'mkdir /foo ; mv /bin/false /foo ; rm -v /foo/false ; ls -l /foo ; rmdir /foo'
+ mkdir /foo
+ mv /bin/false /foo
+ rm -v /foo/false
+ ls -l /foo
total 0
c---------    1 root     root        0,   0 Dec  8 22:51 false
+ rmdir /foo
rmdir: '/foo': Directory not empty
</pre>

**Comment by snnn** [comment]
I can reproduce this bug on Fedora 21.
Docker version: docker-io-1.4.1-8.fc21.x86_64
Kernel version: Linux xxxxx 3.18.3-201.fc21.x86_64 #1 SMP Mon Jan 19 15:59:31 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Comment by ctapan** [comment]
I am not able to reproduce this issue on Ubuntu with older "overlayfs" driver and docker 1.4.1.
3.13.0-44-generic #73-Ubuntu SMP Tue Dec 16 00:22:43 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

Can someone paste the output for following? 
ls -lah /var/lib/docker/overlay/<container-id>/upper/
ls -lah /var/lib/docker/overlay/<container-id>/upper/foo/
ls -lah /var/lib/docker/overlay/<container-id>/upper/bin/

(Maybe with latest "overlay" driver, last directory might be in <container-id>/work/)

Just trying to understand if this is overlay issue or chroot. Most likely will be overlay driver. Sorry I do not have the latest kernel version to try.

**Comment by sgirones** [comment]
Reproduced:
```
$  docker run busybox sh -xec 'mkdir /foo ; mv /bin/false /foo ; rm -v /foo/false ; ls -l /foo ; rmdir /foo'
+ mkdir /foo
+ mv /bin/false /foo
+ rm -v /foo/false
+ ls -l /foo
total 0
c---------    1 root     root        0,   0 Apr 14 11:00 false
+ rmdir /foo
rmdir: '/foo': Directory not empty

$ uname -a
Linux product-builder-1 3.18.11-031811-generic #201504041535 SMP Sat Apr 4 19:37:40 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

$ docker info
Containers: 4
Images: 136
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Kernel Version: 3.18.11-031811-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 2
Total Memory: 23.95 GiB
Name: product-builder-1
ID: GCJZ:YZXU:2V74:NNEE:UUXB:BOC6:LNNC:EELA:C2P6:4DEO:3SJE:N2NZ
WARNING: No swap limit support
```

**Comment by sgirones** [comment]
Reproduced even with 4.0.0. But behavior was slightly different:
```
$ docker run busybox sh -xec 'mkdir /foo ; mv /bin/false /foo ; rm -v /foo/false ; ls -l /foo ; rmdir /foo'
+ mkdir /foo
+ mv /bin/false /foo
+ rm -v /foo/false
+ ls -l /foo
ls: /foo/false: No such file or directory
total 0
```

**Comment by sgirones** [comment]
For reference: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1410480

**Comment by davebirch** [comment]
As an additional piece of potentially useful info, I'm seeing this same issue when using pure lxc with an overlayfs filesystem.

If I delete a file from the lower filesystem, the file is not removed but becomes highlighted as if it's a broken symlink. Trying to read the file results in a "No such file or directory" error.

Current versions:
Linux testcontainer 3.13.0-52-generic #85-Ubuntu SMP Wed Apr 29 16:44:17 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

LXC version 1.0.7

After doing some testing on a couple of different machines, I discovered that this problem occurs on one box I have where the host filesystem is formatted to xfs, but does not occur on my local machine here the / filesystem is formatted to ext4

**Comment by darlingm** [comment]
FWIW, on Fedora 22 (kernel 4.0.4-303.fc22.x86_64), I can't reproduce this bug using overlay and chroot.  (I'm not running lxc or docker.)  After removing "/foo/false", I can `ls -l /foo` without error, and I can `rmdir /foo`.

**Comment by darlingm** [comment]
The (kernel overlayfs documentation)[https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt] does state:

Changes to underlying filesystems

Offline changes, when the overlay is not mounted, are allowed to either the upper or the lower trees.

Changes to the underlying filesystems while part of a mounted overlay filesystem are not allowed.  If the underlying filesystem is changed, the behavior of the overlay is undefined, though it will not result in
a crash or deadlock.

**Comment by smunaut** [comment]
Modifying the lowerfs while there is an overlay mounted over it is not allowed. See /Documentation/filesystems/overlayfs.txt and its "Changes to underlying filesystems" section ...

**Comment by vikstrous** [comment]
This is easy to reproduce with nothing any more special than `mount`.
```
$ mkdir upper lower work merged
$ touch lower/test
$ sudo mount -t overlay overlay -olowerdir=lower,upperdir=upper,workdir=work merged
$ mv merged/test merged/dir/
$ rm merged/dir/test
$ ls -l merged/dir/
/usr/bin/ls: cannot access merged/dir/test: No such file or directory
total 0
c????????? ? ? ? ?            ? test
```

```
$ uname -a
Linux linusputinmusk 4.2.2-1-ARCH #1 SMP PREEMPT Tue Sep 29 22:21:33 CEST 2015 x86_64 GNU/Linux
```

Has this bug been reported to the linux kernel / whoever is working on overlay?

**Comment by vikstrous** [comment]
Oops, I didn't see the link above. Sorry.

It looks like their repro doesn't move the file to a different directory. I don't think this is the same bug. Another hint is that it was marked as fixed and their repro succeeds while ours fails.

So far it looks like it happens only when the directory where the file is moved to and deleted from doesn't exist in the lower layer.

Seems like the issue is that it's creating a whiteout file for a file that never existed in the lower layer. It puts the whiteout file at the last location where the original file was moved to. If that dir doesn't exist in the lower tree the whiteout file shows up in the merged tree as a broken character device file.

This is very clearly a pretty serious bug in overlay. I'm surprised that it was not a bigger problem yet.

**Comment by overplumbum** [comment]
I've some time debugging this issue in kernel. 

I can confirm your idea about state caching — remounting helps because of dentries cache flush. And overlay driver stores metadata exactly in dentry nodes. Looks like VFS swaps nodes with it's meta during rename: thing overlay neither expect nor able to control.

**Comment by xeor** [comment]
Example of weirdness:

    root@2fefd9a8fee7:/home/git/gitlab# ls -l /home/git/gitlab/shared/artifacts/tmp/cache/.gitkeep
    -rw-r--r--. 11 git git 0 Nov 22 07:23 /home/git/gitlab/shared/artifacts/tmp/cache/.gitkeep
    root@2fefd9a8fee7:/home/git/gitlab# rm -f /home/git/gitlab/shared/artifacts/tmp/cache/.gitkeep
    root@2fefd9a8fee7:/home/git/gitlab# ls -l /home/git/gitlab/shared/artifacts/tmp/cache/.gitkeep
    ls: cannot access /home/git/gitlab/shared/artifacts/tmp/cache/.gitkeep: No such file or directory
    root@2fefd9a8fee7:/home/git/gitlab# ls -la /home/git/gitlab/shared/artifacts/tmp/cache/
    ls: cannot access /home/git/gitlab/shared/artifacts/tmp/cache/.gitkeep: No such file or directory
    total 0
    drwxr-xr-x. 1 git git 21 Nov 23 21:56 .
    drwxr-xr-x. 1 git git 18 Nov 23 21:20 ..
    ??????????? ? ?   ?    ?            ? .gitkeep
    root@2fefd9a8fee7:/home/git/gitlab# rm -rf /home/git/gitlab/shared/artifacts/tmp/cache
    rm: cannot remove '/home/git/gitlab/shared/artifacts/tmp/cache': Directory not empty

**Comment by LK4D4** [comment]
@vikstrous Your case works for me on `4.3.0`
However original report now fails for me on `ls` : 
```
+ ls /foo
ls: /foo/false: No such file or directory
```

**Comment by LK4D4** [comment]
I'm pretty sure it should be easy to fix.

**Comment by xeor** [comment]
This problem solved itself for me (as well as for other people in the comments), when going from `xfs` as a storage backend for overlay to `ext4`.

**Comment by vikstrous** [comment]
I opened an issue on the linux kernel bug tracker: https://bugzilla.kernel.org/show_bug.cgi?id=109611

**Comment by mishak87** [comment]
I have same issue on `Linux 1568a0a4bee9 4.2.0-23-generic #28-Ubuntu` using `ext4`.

**Comment by rhvgoyal** [comment]
Proposed a fix upstream.

http://marc.info/?l=linux-fsdevel&m=145252052703010&w=2

**Comment by b-a-t** [comment]
I've experienced the same problem on  Debian 4.3.3-7~bpo8+1 (2016-01-19) kernel with `xfs` storage backstore for overlay FS. Re-formatting it to `ext4` with extra-high i-nodes number seems, solved the problem.

**Comment by koct9i** [comment]
Fixed in linux 4.5 https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=45d11738969633ec07ca35d75d486bf2d8918df6 going to be backported into next 4.4.y and other stable brances. Simple test sequence in commit message.

**Comment by thaJeztah** [comment]
Thanks for the heads up @koct9i

**Comment by mountkin** [comment]
I came across a similar but different problem with overlay+xfs on kernel 4.5.3
After deleting a file in the merged directory, `ls` the file in the merged directory showing all the file attributes as question mark. The file in the upper directory turn out to be a character device.

![3](https://cloud.githubusercontent.com/assets/1496652/15108860/eb49dba0-160a-11e6-9b99-ce97484e2250.png)

**Comment by runcom** [comment]
@mountkin http://article.gmane.org/gmane.linux.file-systems.union/496 there's an attempt to fix that but didn't receive any comments yet

**Comment by koct9i** [comment]
@runcom  Nope, this is different bug. That you're linked was fixed in 4.5 https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=45d11738969633ec07ca35d75d486bf2d8918df6 see my comments above

@mountkin  _this is completely wrong place for reporting linux kernel bugs_
Please submit this into https://bugzilla.kernel.org and/or lkml and related mailing list.

I cannot reproduce this right now in 4.4.8+ext4, I'll try exact setup later.

**Comment by mountkin** [comment]
Thanks a lot @runcom 
The difference between my report and http://article.gmane.org/gmane.linux.file-systems.union/496 is that I don't need to move the file to a sub directory.

@koct9i I'm testing with `xfs`, not `ext4`.  Let me make some noise on the kernel bugzilla then :innocent:

**Comment by koct9i** [comment]
@mountkin  (I haven't found your ticket in bugzilla so I leave it here)
linux 4.6 fails to mount overlayfs if upperdir points to xfs created without ftype=1 and prints this "overlayfs: upper fs needs to support d_type.". I guess your xfs was created without this feature.

related commit https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=45aebeaf4f67468f76bedf62923a576a519a9b68

**Comment by dmcgowan** [comment]
@koct9i thanks for pointing this out. The fix has also been release in 4.4.6 (changelog https://www.kernel.org/pub/linux/kernel/v4.x/ChangeLog-4.4.6). Going to close this issue now.
